### PR TITLE
VM value refactoring

### PR DESCRIPTION
**Pull request by rvanasa**
A starting point for the `Value` -> `Value_` refactor.

### CODE REVIEW

File: crates/motoko/tests/test_dynamic.rs [hunk 1]
```diff
@@ -1,46 +1,71 @@
...
-            .insert(Value::Nat(5.to_biguint().unwrap()), expected.clone());
-        s
-    })));
+    let value = Value::Dynamic(DynamicValue(Box::new(Struct::default())));
 
     let mut core = motoko::vm_types::Core::empty();
-    core.env.insert("value".to_string(), value);
+
+    // TODO: use something like `core.alloc()`
+    let pointer = Pointer(12345);
```
Comment (matthewhammer) [inline crates/motoko/tests/test_dynamic.rs:54]: Yeah, we need a general allocation approach.

Relatedly, there's no "next pointer" field for `Core`, but there should be, and that should be the mechanism for choosing the next pointer name, not the size of the store.

Comment (rvanasa) [reply crates/motoko/tests/test_dynamic.rs]: Done! 

Feel free to change things around, but this should work for now.

File: crates/motoko/src/lib/vm.rs [hunk 2]
```diff
@@ -1284,28 +1258,41 @@ fn stack_cont(core: &mut Core, v: Value) -> Result<Step, Interruption> {
...
                     Ok(Step {})
                 }
                 Value::Null => bang_null(core),
                 _ => Err(Interruption::TypeMismatch),
             },
-            Call1(inst, e2) => match v {
-                Value::Function(cf) => exp_conts(core, FrameCont::Call2(cf, inst), e2),
-                Value::PrimFunction(pf) => exp_conts(core, FrameCont::Call2Prim(pf, inst), e2),
+            Call1(inst, e2) => {
+                let v = core.deref_value(Rc::new(v))?;
```
Comment (matthewhammer) [inline crates/motoko/src/lib/vm.rs:1277]: Introducing a new `Rc::new` here is not "sharing" the original one that was surrounding with `v`.

So, it's not actually doing what we want, if the point is to maximize sharing in the stack frame structure.

Comment (matthewhammer) [reply crates/motoko/src/lib/vm.rs:1277]: Perhaps this is related to the fairly large regression in performance that I think is attributable to this PR?

**Pull request merged**